bitflow/io/sources.py

```diff
@@ -24,13 +33,52 @@ def read_header(marshaller, s, buffer_size=2048):
     return header, over_recv_b
 
 
-class Source(multiprocessing.Process):
+class Source(metaclass=helper.CtrlMethodDecorator):
+
+    def __init__(self, pipeline):
+        self.pipeline = pipeline
+        self.running = multiprocessing.Value('i', 0)
+        self._source = None
+        self.started = False
+
+    def start_and_wait(self):
+        self.start()
+        self.wait()
+
+    def start(self):
+        if not self.started:  # Start only once
+            self.running.value += 1
+            if self._source:
+                self._source.start()
+            else:
+                raise SourceNotDefined("Cannot start. No sample source defined.")
+        self.started = True
+
+    def wait(self):
+        if self._source:
+            self._source.join()
+
+    def stop(self):
+        self.on_close()
+        self.wait()
+
+    def on_close(self):
+        self.running.value = 0
+
+
+class _Source(multiprocessing.Process, metaclass=helper.CtrlMethodDecorator):
 
-    def __init__(self, queue, marshaller, running):
+    def __init__(self, pipeline, marshaller, running):
+        self.pipeline = pipeline
+        self.pipeline_sample_queue_in = pipeline.sample_queue_in
+        self.pipeline_sample_queue_out = pipeline.sample_queue_out
```

Comment:
Is this necessary? Could pipeline just provide a int value with out_samples? Or what else is this queue used for, than just count the number of out samples?

Reply:
I get your point but in the current setup it is required because the destination pipeline of a source does not "know" that its predecessor is a "source". It will read from queue_in and write to queue_out. Without reading the queue_out at source, it will overfill and block. Will also add an join() in on_close to correctly close the queue.

I plan to rework this in the next PR to make it more consistent with other parallel things like batch and fork.